frontend/src/main/scala/bloop/cli/Commands.scala

```diff
@@ -1,25 +1,84 @@
 package bloop.cli
 
+import java.net.InetAddress
+import java.nio.file.Path
+
 import bloop.engine.ExecutionContext
-import caseapp.{ExtraName, HelpMessage, Recurse}
+import caseapp.{CommandName, ExtraName, HelpMessage, Recurse}
 
 object Commands {
-  sealed trait Command {
+
+  /** Represents the most generic command that exists.  */
+  trait Command {
     def cliOptions: CliOptions
   }
 
-  sealed trait CoreCommand extends Command {
+  /** Represents a command that has gone through an initial layer of validation. */
+  sealed trait ValidatedCommand extends Command
+
+  sealed trait ValidatedBsp extends ValidatedCommand
+  case class WindowsLocalBsp(pipeName: String, cliOptions: CliOptions) extends ValidatedBsp
+  case class UnixLocalBsp(socket: Path, cliOptions: CliOptions) extends ValidatedBsp
+  case class TcpBsp(host: InetAddress, port: Int, cliOptions: CliOptions) extends ValidatedBsp
+
+  /** Represents a command that is used by the cli and has no user input validation. */
+  sealed trait RawCommand extends Command
+
+  sealed trait CompilingCommand extends RawCommand {
```

Comment:
`CompilingCommand` makes it sound a little bit like we're pushing command dependencies inside the hierarchy, but I really can't find a better name.

Reply:
With this name I wanted to express the fact that all these commands can trigger compilation.